### PR TITLE
Cleaning up block processing logic

### PR DESCRIPTION
**Pull request by nzpr**
There is two filter cases missing in dependencies set construction - it should not include blocks that are being processed by Casper at the moment or enqueued to Casper.

This PR turned into some bigger attempt to refactor, bugfix and improve how block processing is handled.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Review by tgrospic — APPROVED** [review]
LGTM

**Review by zsluedem — APPROVED** [review]
LGTM

**Comment by nzpr** [comment]
bors merge

**Comment by nzpr** [comment]
bors merge-

**Comment by 9rb** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/Validate.scala
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: Oh, I guess I'm doing it wrong here... this way block will use equivocation as a valid dependency. But I don't want it to be requested again and again.

Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala]: Fix https://github.com/rchain/rchain/pull/3175/commits/65e0277df70a4c70d2a2c2864332548e77595455

**Comment by nzpr** [comment]
The last commit (https://github.com/rchain/rchain/pull/3175/commits/eb6abd6495da1e4feed276108eda3eb07e21fdcf) might fix the problem which @gsj5 seen with block propagation issues. I don't get why I made it `Set` first place and not `Queue` - order does matter here, it should be FIFO.

**Comment by nzpr** [comment]
Sorry for making lots of changes after approval, but I discovered some other flaws that should be addressed. I believe now its complete and ready for review.

File: casper/src/main/scala/coop/rchain/casper/Casper.scala
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/Casper.scala:52]: I believe that if we decide to enable fair dispatcher back - this is is the queue on which it should be done.

**Comment by nzpr** [comment]
This change is too invasive for dev, and should be better done with block merge. To not mess up with rebase, <https://github.com/rchain/rchain/pull/3179> simple change is created.

**Pull request closed without merging**
